tests_and_analysis/test/script_tests/test_dispersion.py

```diff
@@ -1,40 +1,113 @@
+    [cahgo2_fc_file, '-u=1/cm', '--e-min=200'],
+    [cahgo2_fc_file, '--e-min=30', '--e-max=100'],
+    [cahgo2_fc_file, '--length-unit=bohr', '--q-distance=0.04'],
+    [cahgo2_fc_file, '--q-distance=0.02'],
+    [cahgo2_fc_file, '--asr'],
+    [cahgo2_fc_file, '--asr=realspace'],
+    [lzo_fc_file],
+    [nacl_fc_file],
+    [quartz_phonon_file],
+    [quartz_phonon_file, '--btol=1000'],
+    [nacl_phonon_file],
+    [nacl_phonon_hdf5_file]]
+
+
 @pytest.mark.integration
 class TestRegression:
 
     @pytest.fixture
     def inject_mocks(self, mocker):
         # Prevent calls to show so we can get the current figure using
         # gcf()
-        mocker.patch("matplotlib.pyplot.show")
+        mocker.patch('matplotlib.pyplot.show')
         mocker.resetall()
 
     def teardown_method(self):
         # Ensure figures are closed
         matplotlib.pyplot.close('all')
 
-    @pytest.mark.parametrize("dispersion_args", get_dispersion_params())
+    @pytest.mark.parametrize('dispersion_args', disp_params)
     def test_plots_produce_expected_xydata(
             self, inject_mocks, dispersion_args):
-        euphonic.cli.dispersion.main([get_phonon_file()] + dispersion_args)
+        euphonic.cli.dispersion.main(dispersion_args)
 
         lines = matplotlib.pyplot.gcf().axes[0].lines
 
-        with open(get_dispersion_data_file()) as disp_json_file:
-            expected_lines = json.load(
-                disp_json_file)[" ".join(dispersion_args)]
+        with open(disp_output_file, 'r') as f:
+            expected_lines = json.load(f)[args_to_key(dispersion_args)]
+        atol = sys.float_info.epsilon
+        # Increase tolerance if asr present - can give slightly
+        # different results with different libs
+        if any(['--asr' in arg for arg in dispersion_args]):
+            atol = 1.5e-6
         for index, line in enumerate(lines):
             npt.assert_allclose(
                 line.get_xydata().T, np.array(expected_lines[index]),
-                atol=sys.float_info.epsilon)
+                atol=atol)
+
+    @pytest.mark.parametrize('dispersion_args', [
+        [quartz_phonon_file, '--save-to'],
+        [quartz_phonon_file, '-s']])
+    def test_plot_save_to_file(self, inject_mocks, tmpdir, dispersion_args):
+        output_file = str(tmpdir.join('test.png'))
+        euphonic.cli.dispersion.main(dispersion_args + [output_file])
+        assert os.path.exists(output_file)
+
+    @pytest.mark.parametrize('dispersion_args', [
+        [os.path.join(get_data_path(), 'crystal', 'crystal_LZO.json')],
+        [os.path.join(get_data_path(), 'force_constants', 'NaCl',
+                      'FORCE_CONSTANTS')]])
+    def test_invalid_file_raises_value_error(self, dispersion_args):
+        with pytest.raises(ValueError):
+            euphonic.cli.dispersion.main(dispersion_args)
+
+
+@patch('matplotlib.pyplot.show')
+@pytest.mark.skip(reason='Only run if you want to regenerate the test data')
```

Comment:
How confident are we that this cannot be accidentally run with some kind of well-meaning `--seriously-run-all-tests` option to pytest?

Reply:
I had a look, and it seems slightly convoluted to ignore the skip markers (at least in all the Pytest versions released so far) https://stackoverflow.com/questions/56078593/how-to-disable-skipping-a-test-in-pytest-without-modifying-the-code so its probably safe to say it can't be done by accident. If Pytest release a new argument that allows this we may have to rethink it...

Reply:
Ok! If it's not a built-in feature then I'm not too worried.